python/letsql/common/caching.py

```diff
@@ -155,3 +163,33 @@ def _put(self, key, value):
 
     def _drop(self, key):
         self.source.drop_table(key)
+
+
+@frozen
+class SnapshotStorage(SourceStorage):
+    def get_key(self, expr: ir.Expr):
+        typs = map(type, expr.ls.backends)
+        with patch_normalize_token(*typs, f=self.normalize_backend):
+            with patch_normalize_token(
+                ops.DatabaseTable, f=self.normalize_databasetable
+            ):
+                tokenized = dask.base.tokenize(expr)
+                return "-".join(("snapshot", tokenized))
+
+    @staticmethod
+    def normalize_backend(con):
+        name = con.name
+        if name in ("pandas", "duckdb", "datafusion"):
+            return (name, None)
+        else:
+            return normalize_backend(con)
```

Comment:
If the backend name is `postgres,` wouldn't this method reach `RecursionError`?

Reply:
the `normalize_backend` it accesses is the one imported, `self.normalize_backend` **would** result in a `RecursionError`